---
title: "Mouse map grid"
author: "Karl Broman"
date: "3 Nov 2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(results="hide", warning=FALSE, message=FALSE)
options(width=110)
```

This analysis serves to do three things:

- Shift the
  [Liu et al (2014)](https://doi.org/10.1534/genetics.114.161653)
  mouse genetic map so that 0 cM corresponds to 0 Mbp.

- Calculate a grid of markers across the genome.

- Get interpolated cM positions for the markers on the GigaMUGA array.


I'll use the [R/qtl2scan](https://github.com/rqtl/qtl2scan) package to
do interpolations and
[R/qtl2convert](https://github.com/rqtl/qtl2convert) to do some data
manipulation. I'll also use a few other packages below
([rvest](https://github.com/hadley/rvest),
[AnnotationHub](http://bioconductor.org/packages/release/bioc/html/AnnotationHub.html),
[devtools](https://github.com/hadley/devtools)).


```{r load_libraries}
library(qtl2scan)
library(qtl2convert)
```

### Download and prepare the maps

I first need to download the maps from
[Liu et al (2014)](https://doi.org/10.1534/genetics.114.161653) and do
a bit of preparations.

I'll first ensure that I've got directories to contain the files I
download plus those I create.

```{r make_directories}
in_dir <- "files"
out_dir <- "results"
for(dir in c(in_dir, out_dir)) {
    if(!dir.exists(dir)) dir.create(dir)
}
```

Now I'll download
and unzip the [Liu et al (2014)](https://doi.org/10.1534/genetics.114.161653).
The file is about 8 Mbp, because it contains the actual data.

```{r download_liu_map}
url <- "http://cgd.jax.org/mousemapconverter/G2F1.anchor.maps.zip"
file <- file.path(in_dir, basename(url))
if(!file.exists(file)) {
    download.file(url, file)
}
unzip(file, exdir=in_dir)
```

I'll read in the sex-averaged, female, and male genetic maps and get
them aligned. Strangely, they have different numbers of positions.
I'm also going to give each row a name that is like `1_3036178` for
chromosome 1 at 3,036,178 bp.

```{r read_maps}
files <- c("avg.map.csv", "female.map.csv", "male.map.csv")
liu_map <- lapply(files, function(f) read.csv(file.path(in_dir, f)))
liu_map <- lapply(liu_map, function(m) { cbind(m, marker=paste(m$Chr, m$Pos, sep="_")) })
```

I'm going to split them into lists, interpolate positions in the
female and male maps so that they are all the same length, and then
combine them into a single data frame. Note that the male map doesn't
include the X chromosome

```{r align_maps}
liu_pmap_list <- lapply(liu_map, map_df_to_list, chr_column="Chr", pos_column="Pos")
liu_gmap_list <- lapply(liu_map, map_df_to_list, chr_column="Chr", pos_column="cM")
liu_gmap_list[[2]] <- interp_map(liu_pmap_list[[1]], liu_pmap_list[[2]], liu_gmap_list[[2]])
liu_gmap_list[[3]] <- interp_map(liu_pmap_list[[1]][1:19], liu_pmap_list[[3]], liu_gmap_list[[3]])
```

Now I'll combine everything into one data frame and make some slight
corrections (change column names and change chr 20 to chr X).

```{r combine_maps}
liu_map <- cbind(liu_map[[1]],
                 cM_female = map_list_to_df(liu_gmap_list[[2]], pos_column="cM_female")[,"cM_female"],
                 cM_male = c(map_list_to_df(liu_gmap_list[[3]], pos_column="cM_male")[,"cM_male"],
                             rep(NA, sum(liu_map[[1]]$Chr==20))))
colnames(liu_map)[1:2] <- c("chr", "bp")
liu_map$chr[liu_map$chr==20] <- "X"
liu_map$chr <- factor(liu_map$chr, c(1:19,"X"))
liu_map <- liu_map[,c(4,1:3,5,6)]
```

### Get the chromosome lengths

Now we want to anchor the maps at 0 cM. I also want to anchor them at
the telomeres. So I first need to get the chromosome lengths.
I couldn't find the build37 lengths, so I'll start with
the build38 ones, available at <https://www.ncbi.nlm.nih.gov/grc/mouse/data">.

I use [rvest](https://github.com/hadley/rvest) to scrape the
chromosome lengths from the web page.

```{r build38_chr_lengths}
library(rvest)
url <- "https://www.ncbi.nlm.nih.gov/grc/mouse/data"
file <- file.path(in_dir, "grc_mouse_genome_assembly.html")
if(!file.exists(file)) download.file(url, file)
z <- read_html(file)
tab <- html_nodes(z, css="table.ui-ncbigrid")
tab <- html_table(tab)[[1]]
mm10_L <- as.numeric(sapply(strsplit(tab[,2], ","), paste, collapse=""))[1:20]
names(mm10_L) <- c(1:19,"X")
```

But even adding the 3 Mbp padding at the start, chromosome 7 is
shorter than what we have in the Liu et al. maps.

```{r map_lengths, results="markup"}
maxes <- rbind(build38=(mm10_L+3e6) , Liu=tapply(liu_map$bp, liu_map$chr, max))
rbind(maxes, diff=maxes[1,] - maxes[2,])
```

So I'll use
[AnnotationHub](http://bioconductor.org/packages/release/bioc/html/AnnotationHub.html)
to get build 37 (mm9). I'm grabbing the cytobands which seem to be the
smallest thing that will give me full lengths.

```{r annotation_hub_lengths}
file <- file.path(in_dir, "ah51379.rds")
if(file.exists(file)) {
    zz <- readRDS(file)
} else {
    library(AnnotationHub)
    ah <- AnnotationHub()
    z <- query(ah, "mm9")
    zz <- z[["AH53179"]] # cytoband track
    zz <- as.data.frame(zz)
    saveRDS(zz, file)
}
mm9_L <- tapply(zz$end, zz$seqnames, max)
mm9_L <- setNames(mm9_L, substr(names(mm9_L), 4, nchar(mm9_L)))[1:20]
```

Here's a comparison of the maximum positions, vs the Liu et al. map:

```{r map_lengths_mm9, results="markup"}
maxes <- rbind(build38=(mm9_L) , Liu=tapply(liu_map$bp, liu_map$chr, max))
rbind(maxes, diff=maxes[1,] - maxes[2,])
```

Are all longer than Liu map? `r all(mm9_L > tapply(liu_map$bp, liu_map$chr, max))`.
`r ifelse(all(mm9_L > tapply(liu_map$bp, liu_map$chr, max)), "Yay!", "")`


### Shift the maps

Shift the maps so 0 cM == 0 Mbp; also put cap at the telomere.


### Interpolated positions for the GigaMUGA markers

Get interpolated cM positions for the GigaMUGA markers


### Genome grid in cM

Calculate a grid along the genetic map, and find the corresponding
bp positions.


### RNAseq gene positions in cM


### Save stuff to files



### Session Info

```{r session_info, results="markup"}
devtools::session_info()
```
